File: core/src/main/scala/tripaint/effects/ScrambleEffect.scala
package tripaint.effects

import tripaint.coords.GridCoords
import tripaint.grid.ImageGrid

import scala.util.Random

object ScrambleEffect extends Effect {
  def name: String = "Scramble"

  override def action(images: Seq[GridCoords], grid: ImageGrid): Unit = {
    for (imageCoords <- images) {
      val image = grid(imageCoords).get.storage
      val allPixels = image.allPixels

      val transform = allPixels.zip(new Random().shuffle(allPixels).map(image.getColor))

      for ((from, col) <- transform) {
        image.setColor(from, col)
      }
    }
  }
}

File: core/src/main/kotlin/tripaint/effects/ScrambleEffect.kt
package tripaint.effects

import tripaint.coords.GridCoords
import tripaint.grid.ImageGrid
import kotlin.random.Random

object ScrambleEffect : Effect {
    override fun name(): String = "Scramble"

    override fun action(images: List<GridCoords>, grid: ImageGrid) {
        for (imageCoords in images) {
            val image = grid.apply(imageCoords)!!.storage
            val allPixels = image.allPixels()

            val transform = allPixels.zip(Random.shuffled(allPixels).map { image.getColor(it) })

            for ((from, col) in transform) {
                image.setColor(from, col)
            }
        }
    }

    private fun <T> Random.shuffled(xs: List<T>): List<T> {
        val buf = xs.toMutableList()

        fun swap(i1: Int, i2: Int) {
            val tmp = buf[i1]
            buf[i1] = buf[i2]
            buf[i2] = tmp
        }

        for (n in buf.size downTo 2) {
            val k = this.nextInt(n)
            swap(n - 1, k)
        }

        return buf.toList()
    }
}
